### PR TITLE
OpenJDK 21 support

### LINKED ISSUE #28517: Java 21 support
### Description

It'd be good to support the OpenJDK 21, as we want to stick to the latest LTS OpenJDKs. 

The lifecycle of the Java 17 and Java 21 must be properly assessed. 

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### LINKED ISSUE #29968: x509 SAN UPN other name is not handled in JDK 21
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

When dealing with the [support for JDK 21](https://github.com/keycloak/keycloak/pull/28518), I've encountered an issue with the x509 Subject Alternative Name (SAN) User Principal Name (`otherName`) in certificates. 

It seems the attribute is not properly handled when using JDK 21. It works properly with JDK 17.

To check the SAN props of certificate with `openssl x509 -noout -text -in test-user-san@localhost.cert.pem`:

```
X509v3 Subject Alternative Name: 
                email:test-user-altmail@localhost, IP Address:192.168.7.1, DNS:www.example-test.com, URI:http://www.example-test.com, othername: 1.2.3.4::my_test_user, othername: UPN::test_upn_name@localhost
```

It was caught by test `X509BrowserLoginSubjectAltNameTest#loginAsUserFromCertSANUpn` and log with the failed test:

```
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,578 INFO  [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] [X509BrowserLoginSubjectAltNameTest] loginAsUserFromCertSANUpn() STARTED
X509BrowserLoginSubjectAltNameTest ++ May 29, 2024 8:50:02 AM org.jboss.arquillian.drone.webdriver.factory.HtmlUnitDriverFactory setClientOptions
X509BrowserLoginSubjectAltNameTest ++ INFO: Setting HtmlDriver web client options: {htmlunitCssEnabled=false, htmlunitHistoryPageCacheLimit=1}
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,588 INFO  [org.keycloak.testsuite.drone.KeycloakDronePostSetup] Configuring driver settings. implicitWait=2000, pageLoadTimeout=10000
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,588 INFO  [org.keycloak.testsuite.drone.KeycloakDronePostSetup] Check HtmlUnit driver TLS settings
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,588 INFO  [org.keycloak.testsuite.drone.KeycloakDronePostSetup] Keystore '/home/runner/work/keycloak/keycloak/testsuite/integration-arquillian/tests/base/target/containers/auth-server-quarkus/conf/certs/clients/test-user-san-cert-test-user-key@localhost.p12', password 'password', type 'pkcs12'
X509BrowserLoginSubjectAltNameTest ++ May 29, 2024 8:50:02 AM org.jboss.arquillian.drone.webdriver.factory.HtmlUnitDriverFactory setClientOptions
X509BrowserLoginSubjectAltNameTest ++ INFO: Setting HtmlDriver web client options: {htmlunitCssEnabled=false, htmlunitHistoryPageCacheLimit=1}
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,614 INFO  [org.keycloak.testsuite.drone.KeycloakDronePostSetup] Configuring driver settings. implicitWait=2000, pageLoadTimeout=10000
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,615 INFO  [org.keycloak.testsuite.drone.KeycloakDronePostSetup] Check HtmlUnit driver TLS settings
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,615 INFO  [org.keycloak.testsuite.drone.KeycloakDronePostSetup] Keystore '/home/runner/work/keycloak/keycloak/testsuite/integration-arquillian/tests/base/target/containers/auth-server-quarkus/conf/certs/clients/test-user-san-cert-test-user-key@localhost.p12', password 'password', type 'pkcs12'
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,656 DEBUG [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] loading test realms
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,656 DEBUG [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] Adding test realm for import from testrealm.json
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,657 INFO  [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] importing test realms
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,657 DEBUG [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] --importing realm: test
X509BrowserLoginSubjectAltNameTest ++ 2024-05-29 08:50:04,299 ERROR [org.keycloak.services] (executor-thread-2) [X509ClientCertificateAuthenticator:authenticate] Exception: class java.lang.String cannot be cast to class [B (java.lang.String and [B are in module java.base of loader 'bootstrap')
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,310 INFO  [org.keycloak.testsuite.util.URLUtils] starting navigation
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,310 INFO  [org.keycloak.testsuite.util.URLUtils] current URL:  https://localhost:8543/auth/realms/test/protocol/openid-connect/auth?response_type=code&client_id=test-app&redirect_uri=https%3A%2F%2Flocalhost%3A8543%2Fauth%2Frealms%2Fmaster%2Fapp%2Fauth&state=693b74d2-2a3a-4a53-bde5-a0e0acfa1bed&scope=openid
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,310 INFO  [org.keycloak.testsuite.util.URLUtils] navigating to https://localhost:8543/auth/realms/test/testing/blank
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,316 INFO  [org.keycloak.testsuite.util.URLUtils] new current URL:  https://localhost:8543/auth/realms/test/testing/blank
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,316 INFO  [org.keycloak.testsuite.util.URLUtils] navigation complete
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,316 INFO  [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] deleting cookies in 'test' realm
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,316 INFO  [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] removing test realms after test method
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,316 INFO  [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] removing realm: test
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,514 ERROR [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] [X509BrowserLoginSubjectAltNameTest] loginAsUserFromCertSANUpn() FAILED
...
...
Error: OR]   Run 1: X509BrowserLoginSubjectAltNameTest.loginAsUserFromCertSANUpn:57->AbstractX509AuthenticationTest.x509BrowserLogin:571 » NoSuchElement Unable to locate element with ID: 'certificate_subjectDN'
-- For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#no-such-element-exception
-- Build info: version: '4.21.0', revision: '79ed462ef4'
-- System info: os.name: 'Linux', os.arch: 'amd64', os.version: '6.5.0-1021-azure', java.version: '21.0.3'
-- Driver info: driver.version: HtmlUnitDriver
```

https://github.com/keycloak/keycloak/actions/runs/9271211106/job/25541538448?pr=28518

It seems there's certain improper type casting based on the log (not shown on local machine):
```
X509BrowserLoginSubjectAltNameTest ++ 2024-05-29 08:50:04,299 ERROR [org.keycloak.services] (executor-thread-2) [X509ClientCertificateAuthenticator:authenticate] Exception: class java.lang.String cannot be cast to class [B (java.lang.String and [B are in module java.base of loader 'bootstrap')
```  

Even though there were recent changes(https://github.com/keycloak/keycloak/commit/76a6733f0a0edf25f3e75724b0cf708f0284e19f#diff-9e112d961710fdb5fb2821f23a9c732eab5d5fc4e4f6c89d9157d3eb46597e43R47) in the test class, it should reflect the situation same as before as the certificates are identical (checked by the openssl tool).

In normal situations, the page should redirect to the `login-x509-info.ftl`, but with the JDK 21 and the specific test case, the page is redirected to the login screen and requires authentication.

@mposolda @rmartinc Could you please check it? 

### Version

main

### Regression

- [ ] The issue is a regression

### Expected behavior

Test pass

### Actual behavior

Test does not pass

### How to Reproduce?

1. Checkout to https://github.com/keycloak/keycloak/pull/28518 branch in order to properly execute these tests
2. Change Java version to 21 (_mvn needs to reflect it as well_)
3. `mvn clean install -Dtest=X509BrowserLoginSubjectAltNameTest#loginAsUserFromCertSANUpn -f testsuite/integration-arquillian/tests/base/pom.xml`

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by mabartos**
Closes #28517
Closes #29968

# Discussion

We need to assess the lifecycle of the currently supported OpenJDK 17 with the OpenJDK 21.
We may face some obstacles, which need a discussion. 

## Some considerations

- We'd like to stick to the latest LTS OpenJDK versions
- If we support OpenJDK 17 and OpenJDK 21 simultaneously, testing would need to be doubled = **NOK**
- Or we can test OpenJDK 17 only during nightly builds
- We can deprecate OpenJDK 17 and claim the testing of it will be limited
- If deprecated for Keycloak 25 and removed in **Keycloak 26**, it'd be a bad move for RHBK 26 (support for JDK 17 should be set - AFAIK) = **NOK**
- If deprecated for Keycloak 25 and removed in **Keycloak 27**, the OpenJDK 17 would be deprecated for RHBK 26 and removed in RHBK ~28 (potentially) = **OK**
- Java Distribution IT tests check OpenJDK 17 ~~and the latest JDK (OpenJDK 22)~~ [Invalid CEN header (invalid zip64 extra data field size](https://bugs.openjdk.org/browse/JDK-8313765) is thrown due to the additional ZIP64 validation)

## Summary
I'd suggest the following:

- OpenJDK 21 is the default (users can still use JDK17 and JDK17-based extensions)
- Deprecate OpenJDK 17 for Keycloak 25
- Remove OpenJDK 17 in Keycloak ~27
- Test OpenJDK 17 only as part of Java Distribution IT tests

---------------

@stianst @vmuzikar @ahus1 @miquelsi 
What are the requirements for the JDK support, and what are your thoughts on this?

### CODE REVIEW

**Comment by ahus1** [comment]
@pruivo - could please comment if Infinispan 14.x officially support OpenJDK 21? This is more of a yes/no question. If it is no, we would need to discuss if/where/how we would request such a compatibility. Thanks!

**Comment by pruivo** [comment]
@ahus1 simple answer:
* Infinispan 14 supports OpenJDK 21 and virtual threads (disabled by default, needs to be explicitly enabled via Java properties)
* RHDG 8.4 (based in ISPN 14) will not support OpenJDK 21 and does not have virtual threads support.

**Comment by tristantarrant** [comment]
> @ahus1 simple answer:
>     * RHDG 8.4 (based in ISPN 14) will not support OpenJDK 21 and does not have virtual threads support.

Let me clarify: RHDG 8.4 not supporting JDK 21 is merely a QE thing. It will work on JDK 21.
vthreads support is a different matter. While upstream has support for them, they are disabled by default.

**Comment by iot2198** [comment]
I have a question regarding support with jdk 21.
I have few SPI that are being injected into keycloak image, these SPI are build on java 21 which eventually leads to failure as keycloak has jdk17 in built support.
How do we overcome this @mabartos @tristantarrant @ahus1

**Comment by shawkins** [comment]
@iot2198 if using maven, you can set the compile target to 17, or to be extra safe set the release https://maven.apache.org/plugins/maven-compiler-plugin/examples/set-compiler-release.html

**Comment by iot2198** [comment]
for gradle we would have to manually install jdk 21 in it ?
i am making a custom image so it would require me to download it and install it, something along these lines:

`FROM quay.io/keycloak/keycloak:24.0.4

# Install JDK 21 manually
USER root
RUN curl -L -o openjdk-21_linux-x64_bin.tar.gz https://download.java.net/java/GA/jdk21/21/GPL/openjdk-21_linux-x64_bin.tar.gz && \
    mkdir /opt/jdk-21 && \
    tar -xvf openjdk-21_linux-x64_bin.tar.gz -C /opt/jdk-21 --strip-components=1 && \
    rm openjdk-21_linux-x64_bin.tar.gz

# Set JAVA_HOME to JDK 21 and update PATH
ENV JAVA_HOME=/opt/jdk-21
ENV PATH=$JAVA_HOME/bin:$PATH`

**Comment by pruivo** [comment]
@iot2198 simple, or you compile your code with JDK17 compatibility or create a fresh image of Keycloak with JDK 21 installed instead of using the official one :+1:

**Comment by mabartos** [comment]
Failed tests were related to certain instabilities -> restarted.

**Comment by mabartos** [comment]
@stianst @vmuzikar @ahus1 @keycloak/cloud-native-maintainers

Could you please check it? Thanks!

File: quarkus/container/Dockerfile
Comment (vmuzikar) [inline quarkus/container/Dockerfile:1]: Should we make the switch to Java 21 in the Operator Dockerfile too?

File: docs/documentation/release_notes/topics/25_0_0.adoc
Comment (vmuzikar) [inline docs/documentation/release_notes/topics/25_0_0.adoc:1]: With regards to docs, IMHO we should also update the Getting Started guides that instruct users to install Java 17.

File: docs/documentation/release_notes/topics/25_0_0.adoc
Comment (vmuzikar) [inline docs/documentation/release_notes/topics/25_0_0.adoc:91]:
```suggestion
= Java 17 support is deprecated
```

File: docs/documentation/release_notes/topics/25_0_0.adoc
Comment (vmuzikar) [inline docs/documentation/release_notes/topics/25_0_0.adoc:93]:
```suggestion
OpenJDK 17 support is deprecated in {project_name}, and will be removed in a following release in favor of OpenJDK 21.
```

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@mabartos Nice work, thank you!

Added some comments.

Also, I believe we need a review from someone from core and possibly @ASzc as well (due to ant changes).

**Comment by mabartos** [comment]
@vmuzikar Oops, not sure why I missed the docs, thanks! Added a new commit.

**Review by vmuzikar — DISMISSED** [review]
@mabartos Thanks for the update. LGTM.

We might want to consider having the Java version as a variable in the docs. But that's something for the future. :)

**Review by miquelsi — APPROVED** [review]
Great job @mabartos, thanks!

**Comment by mposolda** [comment]
@vmuzikar @mabartos @stianst I am not sure if we want to officially deprecate Java 17 in release notes considering FIPS? Not 100% sure if we need to use "RHEL ELS" image for the FIPS 140 support, which might be based on RHEL 8.6. But RHEL 8 seems to support Java 21 from 8.9 based on: https://access.redhat.com/articles/1299013 . Or do we still deprecate Java 17 and then re-evaluate in Keycloak 26?

**Review by mposolda — DISMISSED** [review]

**Review by vmuzikar — DISMISSED** [review]

**Comment by mposolda** [comment]
@mabartos I think it is normal DCO usually fails in this case with 2 commits, but if we squash commits during merge, it might be ok

**Comment by vmuzikar** [comment]
I've overridden the DCO check, it was a false positive.

Also, some tests failed, I tried re-running them.

**Review by mposolda — DISMISSED** [review]

File: docs/documentation/release_notes/topics/25_0_0.adoc
Comment (stianst) [inline docs/documentation/release_notes/topics/25_0_0.adoc:94]:
```suggestion
```

Comment (mabartos) [reply docs/documentation/release_notes/topics/25_0_0.adoc:94]: ok

**Comment by mabartos** [comment]
We're waiting for resolving WebAuthn tests by: https://github.com/keycloak/keycloak/pull/30027

**Comment by mabartos** [comment]
@mposolda @vmuzikar Addressed Stian's comment and rebased. WebAuthn tests should be fine now.

**Review by vmuzikar — APPROVED** [review]

**Comment by mabartos** [comment]
@vmuzikar I think we're ready to go, right?

The only failure is related to DCO:
```
Commit sha: c5d01ff, Author: Marek Posolda, Committer: Martin Bartoš; Expected "Marek Posolda mposolda@gmail.com", but got "mposolda mposolda@gmail.com".
```
But AFAIK, we accept 'username' as the DCO, so I set DCO to pass.

**Comment by LaughDonor** [comment]
Naive question here, would it be simpler to use the OpenJDK21 container version of ubi9 here? https://catalog.redhat.com/software/containers/ubi9/openjdk-21-runtime/6501ce769a0d86945c422d5f

This is in reference to the Dockerfiles:
https://github.com/keycloak/keycloak/blob/7eef7697e6340ea367356293374b28de5f181471/operator/Dockerfile#L1

https://github.com/keycloak/keycloak/blob/7eef7697e6340ea367356293374b28de5f181471/quarkus/container/Dockerfile#L1

Please let me know if I should create a new issue.

**Comment by mabartos** [comment]
@LaughDonor Thanks for the additional observation! 

It'd be great if you could create an enhancement issue where you describe this case and the benefits that it can bring to Keycloak. Then, we can discuss this further with our wider team.

**Pull request merged**
